Mathlib/Combinatorics/Colex.lean

```diff
@@ -1,423 +1,414 @@
-    ((∑ i in A, 2 ^ i) < ∑ i in B, 2 ^ i) ↔ A.toColex < B.toColex := by
-  have z : ∀ A B : Finset ℕ, A.toColex < B.toColex → ∑ i in A, 2 ^ i < ∑ i in B, 2 ^ i := by
-    intro A B
-    rw [← sdiff_lt_sdiff_iff_lt, Colex.lt_def]
-    rintro ⟨k, z, kA, kB⟩
-    rw [← sdiff_union_inter A B]
-    conv_rhs => rw [← sdiff_union_inter B A]
-    rw [sum_union (disjoint_sdiff_inter _ _), sum_union (disjoint_sdiff_inter _ _), inter_comm,
-      add_lt_add_iff_right]
-    apply lt_of_lt_of_le (@Nat.sum_two_pow_lt k (A \ B) _)
-    · apply single_le_sum (fun _ _ => Nat.zero_le _) kB
-    intro x hx
-    apply lt_of_le_of_ne (le_of_not_lt _)
-    · apply ne_of_mem_of_not_mem hx kA
-    -- Porting note: `intro` required because `apply` behaves differently
-    intro kx
-    have := (z kx).1 hx
-    rw [mem_sdiff] at this hx
-    exact hx.2 this.1
-  refine'
-    ⟨fun h => (lt_trichotomy A B).resolve_right fun h₁ => h₁.elim _ (not_lt_of_gt h ∘ z _ _), z A B⟩
-  rintro rfl
-  apply irrefl _ h
-#align colex.sum_two_pow_lt_iff_lt Colex.sum_two_pow_lt_iff_lt
-
-/-- For subsets of ℕ, we can show that colex is equivalent to binary. -/
-theorem sum_two_pow_le_iff_lt (A B : Finset ℕ) :
-    ((∑ i in A, 2 ^ i) ≤ ∑ i in B, 2 ^ i) ↔ A.toColex ≤ B.toColex := by
-  rw [le_iff_le_iff_lt_iff_lt, sum_two_pow_lt_iff_lt]
-#align colex.sum_two_pow_le_iff_lt Colex.sum_two_pow_le_iff_lt
+lemma mem_initSeg : t ∈ initSeg s ↔ s.card = t.card ∧ toColex t ≤ toColex s := by simp [initSeg]
+
+lemma mem_initSeg_self : s ∈ initSeg s := by simp
+@[simp] lemma initSeg_nonempty : (initSeg s).Nonempty := ⟨s, mem_initSeg_self⟩
+
+lemma isInitSeg_initSeg : IsInitSeg (initSeg s) s.card := by
+  refine ⟨fun t ht => (mem_initSeg.1 ht).1.symm, fun t₁ t₂ ht₁ ht₂ ↦ mem_initSeg.2 ⟨ht₂.2.symm, ?_⟩⟩
+  rw [mem_initSeg] at ht₁
+  exact ht₂.1.le.trans ht₁.2
+
+lemma IsInitSeg.exists_initSeg (h𝒜 : IsInitSeg 𝒜 r) (h𝒜₀ : 𝒜.Nonempty) :
+    ∃ s : Finset α, s.card = r ∧ 𝒜 = initSeg s := by
+  have hs := sup'_mem (ofColex ⁻¹' 𝒜) (LinearOrder.supClosed _) 𝒜 h𝒜₀ toColex
+    (fun a ha ↦ by simpa using ha)
+  refine' ⟨_, h𝒜.1 hs, _⟩
+  ext t
+  rw [mem_initSeg]
+  refine' ⟨fun p ↦ _, _⟩
+  · rw [h𝒜.1 p, h𝒜.1 hs]
+    exact ⟨rfl, le_sup' _ p⟩
+  rintro ⟨cards, le⟩
+  obtain p | p := le.eq_or_lt
+  · rwa [toColex_inj.1 p]
+  · exact h𝒜.2 hs ⟨p, cards ▸ h𝒜.1 hs⟩
+
+/-- Being a nonempty initial segment of colex if equivalent to being an `initSeg`. -/
+lemma isInitSeg_iff_exists_initSeg :
+    IsInitSeg 𝒜 r ∧ 𝒜.Nonempty ↔ ∃ s : Finset α, s.card = r ∧ 𝒜 = initSeg s := by
+  refine ⟨fun h𝒜 ↦ h𝒜.1.exists_initSeg h𝒜.2, ?_⟩
+  rintro ⟨s, rfl, rfl⟩
+  exact ⟨isInitSeg_initSeg, initSeg_nonempty⟩
 
 end Colex
+
+open Colex
+
+/-!
+### Colex on `ℕ`
+
+The colexicographic order agrees with the order induced by interpreting a set of naturals as a
+binary expansion.
+-/
+
+section Nat
+variable {s t : Finset ℕ}
+
+private lemma aux : StrictMono (fun s : Colex ℕ ↦ ∑ k in ofColex s, 2 ^ k) := by
```

Comment:
I think it might be worth keeping this public, for instance to derive injectivity or if composition of this with other things is useful. In fact I think it would be great to make this a order isomorphism, but that doesn't need to be in this PR.